### PR TITLE
Refactor `v1/search`: include WHERE condition, allow extra columns in projection.

### PR DESCRIPTION
**Pull request by Jeadie**
## 🗣 Description
- New format for `/v1/search` (see below).
- New capabilities. These also work for `document_similarity` tool:
    - filter vector search on `WHERE` conditional
    - Return additional columns from the dataset.

#### Request
```rust
pub struct SearchRequest {
    /// The text to search documents for similarity
    pub text: String,

    /// The datasets to search for similarity. For available datasets, use the 'list_datasets' tool
    #[serde(default)]
    pub datasets: Vec<String>,

    /// Number of documents to return for each dataset
    #[serde(default = "default_limit")]
    pub limit: usize,

    /// An SQL filter predicate to apply. Format: 'WHERE <where_cond>'.
    #[serde(rename = "where", default)]
    pub where_cond: Option<String>,

    /// Additional columns to return from the dataset.
    #[serde(default)]
    pub additional_columns: Vec<String>,
}
```
#### Response
```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
struct SearchResponse {
    matches: Vec<Match>,
    duration_ms: u128,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Match {
    value: String,
    dataset: String,
    primary_key: HashMap<String, serde_json::Value>,
    metadata: HashMap<String, serde_json::Value>,
}
```

## 🤔 Concerns
 - There's no great way to convert a `RecordBatch` to a `serde_json::Value`. For now, converting to JSON string, then into a `Value` is the chosen option.

### CODE REVIEW

File: crates/runtime/src/model/tool_use.rs [hunk 1]
```diff
@@ -130,6 +130,10 @@ static DOCUMENT_SIMILARITY_PARAMETERS: Lazy<Value> = Lazy::new(|| {
                 "description": "The datasets to search for similarity. For available datasets, use the 'list_datasets' tool",
```
Comment (lukekim) [inline crates/runtime/src/model/tool_use.rs]: `from` -> `datasets`

File: crates/runtime/src/embeddings/vector_search.rs [hunk 2]
```diff
@@ -87,9 +98,84 @@ impl Display for RetrievalLimit {
...
+}
+
+pub type VectorSearchResult = HashMap<TableReference, VectorSearchTableResult>;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Match {
+    value: String,
+    // score: f64,
+    dataset: String,
+    primary_keys: HashMap<String, serde_json::Value>,
```
Comment (lukekim) [inline crates/runtime/src/embeddings/vector_search.rs]:
```suggestion
    primary_key: HashMap<String, serde_json::Value>,
```

File: crates/runtime/src/embeddings/vector_search.rs [hunk 3]
```diff
@@ -87,9 +98,84 @@ impl Display for RetrievalLimit {
...
+    metadata: HashMap<String, serde_json::Value>,
+}
+
+pub fn table_to_matches(
+    tbl: &TableReference,
+    result: &VectorSearchTableResult,
+) -> Result<Vec<Match>> {
+    let pks: Vec<HashMap<String, serde_json::Value>> = if result
+        .primary_keys
```
Comment (lukekim) [inline crates/runtime/src/embeddings/vector_search.rs:123]:
```suggestion
        .primary_key
```

File: crates/runtime/src/embeddings/vector_search.rs [hunk 4]
```diff
@@ -87,9 +98,84 @@ impl Display for RetrievalLimit {
...
+                vec![]
+            }
+        })
+        .collect();
+
+    Ok(zip(zip(pks, add_cols), values)
+        .map(|((pks, add_cols), value)| Match {
+            value,
+            dataset: tbl.to_string(),
+            primary_keys: pks,
```
Comment (lukekim) [inline crates/runtime/src/embeddings/vector_search.rs]:
```suggestion
            primary_key: pks,
```

File: crates/runtime/src/embeddings/vector_search.rs [hunk 5]
```diff
@@ -44,6 +47,14 @@ pub enum Error {
         source: Box<dyn std::error::Error + Send + Sync>,
     },
 
+    #[snafu(display("Error occurred process Arrow records: {}", source))]
```
Comment (phillipleblanc) [inline crates/runtime/src/embeddings/vector_search.rs]:
```suggestion
    #[snafu(display("Error occurred processing Arrow records: {}", source))]
```

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**